kano_profile/badges.py

```diff
@@ -104,87 +119,153 @@ def calculate_min_current_max_xp():
+        self._all_rules = load_badge_rules()
+        self._calculated_badges = {}
+
+    def _parse_target_variable(self, variable):
+        return variable.split('.')
+
+    def _get_variable(self, app, levels):
+        ret_val = None
+        try:
+            ret_val = self._app_state[app]
+            for k in xrange(len(levels)):
+                ret_val = ret_val[levels[k]]
+        except (TypeError, KeyError) as exc:
+            logger.debug(
+                "Can't find levels {} in app {}, - {}".format(levels, app, exc)
+            )
+            ret_val = None
+        return ret_val
+
+    def _evaluate_rules(self, rules):
+        if 'operation' not in rules:
+            return None
+
+        if rules['operation'] == 'each_greater':
+            achieved = True
+            for target in rules['targets']:
+                app = target[0]
+                attr_list = self._parse_target_variable(target[1])
+                threshold_value = target[2]
+
+                local_value = self._get_variable(app, attr_list)
+                if local_value is None:
+                    achieved = False
+                if attr_list[-1] == 'level' and threshold_value == -1:
+                    threshold_value = self._app_profiles[app]['max_level']
+                    break
+
+                achieved &= local_value >= threshold_value
+
+        elif rules['operation'] == 'sum_greater':
+            acc = 0
+            for target in rules['targets']:
+                app = target[0]
+                attr_list = self._parse_target_variable(target[1])
+
+                local_value = self._get_variable(app, attr_list)
+                if local_value is None:
+                    continue
+
+                acc += float(local_value)
+
+            achieved = acc >= rules['value']
+        else:
+            achieved = None
 
-                    elif rules['operation'] == 'sum_greater':
-                        sum = 0
-                        for target in rules['targets']:
-                            app = target[0]
-                            variable = target[1]
+        return achieved
 
-                            if app not in app_list or variable not in app_state[app]:
-                                continue
+    @staticmethod
+    def is_push_back(item):
+        rules = item[1]
+        return 'push_back' in rules and rules['push_back'] is True
 
-                            sum += float(app_state[app][variable])
+    def do_calculate(self, calculate_only_pushed_back):
+        for category, subcats in self._all_rules.iteritems():
+            for subcat, items in subcats.iteritems():
+                if calculate_only_pushed_back:
+                    filter_fn = itertools.ifilter
+                else:
+                    filter_fn = itertools.ifilterfalse
 
-                        achieved = sum >= rules['value']
+                it = filter_fn(self.is_push_back, items.iteritems())
```

Comment:
I'm not sure if ifilter would know to chain the predicates. Wouldn't it try to apply a `calculate_only_pushed_back and self.is_push_back.__call__(...)`